content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+Cons:
+
+* outside the developers’ usual workflow
+
+* distinguishing between new and existing static-analysis issues was distracting
+
+This resonates with my personal experience — in 2007, as a Build Engineer on big Java project at the first place of job, a consultancy outlet, I have experience creating internal **“Personalized Quality Dashboard”** service.
+
+It was a bunch of static HTML files, that later became a database for a web application, produced as a part of nightly build using Maven that were copied and served from a well known URL inside the company. HTMLs would contain tables of potential software defects, obtained by running multiple existing static analysis tools — [FindBugs](https://github.com/findbugsproject/findbugs), [PMD](https://pmd.github.io/) and [CheckStyle](https://github.com/checkstyle/checkstyle). Each defect was attributed to the latest change in the codebase using “git blame” and “assigned” to a particular engineer who introduced the change.
+
+It’s internal adoption numbers, although driven top-down by the management decision, were *very low* and aligned with the paper — needless to say that not many engineers were motivated enough to go to a separate [http://code-quality.company.com](http://code-quality.company.com) every day, only to find that they are #N by the amount of bugs introduced to the codebase.
+
+Curiously enough, one can see some open source projects like i.e Git going though the similar stage right now i.e [https://larsxschneider.github.io/git-scan](https://larsxschneider.github.io/git-scan/) with contributors introducing language-specific analysis tools to the build profiles and publishing a dashboards with the results.
+
+Despite challenges in adopting such solutions, one can also see companies, i.e [https://scan.coverity.com](https://scan.coverity.com) — a closed-sourced static analysis 
+ solution for Java, C/C++, C#, JavaScript, Ruby and Python [founded in 2006](https://scan.coverity.com/about) jointly with U.S. Department of Homeland Security, being gradually adopted by some OSS projects.
+ 
+Companies building rule-based analysis platforms like [https://lgtm.com](https://lgtm.com), offspring of University of Oxford-based [https://semmle.com](https://semmle.com/) founded in 2007, are following this adoption path. Theri success, in my opinion, can be attributed to the fact that both support “hard” native languages like [C++](https://lgtm.com/blog/how_lgtm_builds_cplusplus).
+
+### 2. 2009 Filing bugs/Fixit
+
+Next attempt introducing static analysis tools for Java, documented in the paper was filing the results of analysis as bugs in the project bug-tracking system. Then, a companywide dedicated effort was made though a “Fixit” week for all engineers to have a time to clean up those issues.
+
+This approach has some advantages:
+
+* it is valid scientific approach as it allows to quantify the results very well: how many of reported issues were actually fixed by developers
```

Comment:
it is **a** ~valid~ proven

Reply:
My concern is that "Proven" in this context does not seem to be used very often and also may rise unnecessary questions like "proven by whom? where is the proof?", etc

Reply:
The problem with "valid" is that it assumes that there is an invalid scientific approach, which is ugly too. How about "good" or "reliable"?